### ISSUE TITLE
Support GNNAsKernel

### ISSUE THREAD

**Issue by LingxiaoShawn** [pyg-team/pytorch_geometric#3664]
## 🚀 Feature

[GNNAsKernel](https://github.com/GNNAsKernel/GNNAsKernel) is a general framework that boosts the expressiveness of any existing message-passing based GNNs. The original implementation supports any layer from PyG with a wrapper around it. Adding the support of GNNAsKernel can greatly improve the usage of message-passing GNNs in real-world. 
### Motivation

* Message passing based GNNs are limited by 1-WL in terms of expressiveness.
* GNN-AK uplifts all these GNNs to be more expressive, with great practical performance and great scalability. 

### Pitch

Adding GNN-AK to PyG with an option to uplift any existing graph convolutional layer in PyG. The current implementation inside [GNNAsKernel](https://github.com/GNNAsKernel/GNNAsKernel) needs a wrapper around layers from PyG. How to remove it or define a general wrapper to make it easy to plug-in?

**Comment by wsad1** [comment]
@LingxiaoShawn  thanks for proposing this. One approach would be a to add this to torch_geometric.nn.models.
Something like `GNN_AK(base_gnn, **kwargs)`. it would have a `train` routine that can probably use the [torch_geometric.loader.ShaDowKHopSample](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.ShaDowKHopSampler) to get sampled subgraphs for training.

**Comment by rusty1s** [comment]
As far as I see, there are four steps involved in adding `GNN-AK` to PyG:
1. Add `data` transformations
2. Add model conversion/wrapper
3. Add sampler part
4. Add an example to showcase its usage

@LingxiaoShawn Can you confirm? I think we should create a PR for each of those components. One thing I'm super interested in is how we can implement the wrapper part. Can you walk us through the current implementation of the wrapper?

**Comment by LingxiaoShawn** [comment]
Hi @rusty1s I think the steps are good in general.  Transformation is separate from model while the model uses additional variables generated from transformation, which has a strict dependence on the transformation. Sampler part is additional feature added into transformation.  

I'm happy to setup a zoom meeting to discuss the detail! Please let me know.

**Comment by LingxiaoShawn** [comment]
> would have a `train` routine that can

Add a model to torch_geometric.nn.models probably is the easiest way. However different from other models that are specific for problem, GNN-AK is a general framework (a bit more general than model?). The sampler you mentioned probably cannot do the job, I suggest to use the transformation layer in the source code. Thanks!

**Comment by rusty1s** [comment]
Sure, we can discuss in Zoom as well (although I'm not sure we can find a date till the end of this year). As an alternative, how about we create a first PR for `data` transformations (feel free to send your version which we can refine). This should give me and @wsad1 a good starting point.

**Comment by LingxiaoShawn** [comment]
Hey @rusty1s yeah it works for me. I think the code in the github is kind of clear for the transformation layer, but do let me know if you need help on rewriting any part!

**Comment by rusty1s** [comment]
Super, feel free to send a PR so we can acknowledge your code base :)

**Comment by Padarn** [comment]
Hey, guys is this currently being incorporated. I think I can do a rewrite/port of @LingxiaoShawn's code for the first part of this if there isn't already someone doing so.

**Comment by rusty1s** [comment]
Yes, we have a first open PR in https://github.com/pyg-team/pytorch_geometric/pull/3926. If you are interested, you can help us reviewing :)

**Comment by Padarn** [comment]
Awesome, looking forward to seeing this available!
